src/lib/interruptible/interruptible.ml

```diff
@@ -11,48 +11,56 @@ module T = struct
     ; d= Deferred.Result.map_error d ~f }
 
   let bind t ~f =
-    match Deferred.peek t.d with
-    | None ->
-        let interruption_signal_to_res =
-          t.interruption_signal >>| fun s -> Error s
-        in
-        let maybe_sig_d =
-          let w : ('a, 's) Deferred.Result.t =
-            Deferred.any [t.d; interruption_signal_to_res]
+    match Deferred.peek t.interruption_signal with
```

Comment:
This behaviour of interruptible is better, but I'm still not sure that it's right. I've opened #6889 to implement (what I consider to be) the correct behaviour.